iocore/utils/PartitionedMap.h

```diff
@@ -0,0 +1,287 @@
+#ifndef _PartitionedMap_h
+#define _PartitionedMap_h
+
+#include "stdint.h"
+#include <cstddef>
+#include <mutex>
+#include <unordered_map>
+#include <atomic>
+#include <thread>
+
+#include "ts/ink_assert.h"
+
+using Mutex     = std::recursive_mutex;
+using LockGuard = std::lock_guard<Mutex>;
+
+inline bool
+threadHasLock(Mutex &m)
+{
+  return pthread_t(m.native_handle()) == pthread_self();
+}
+
+/**
+ * @brief Low overhead implementation of a recursive read write lock.
+ *
+ * This is does deadlock if you try to acquire both a read and then a write
+ * lock on the same thread.
+ *
+ * I copied the API from std::shared_mutex, so this can be used with std::shared_lock,
+ * std::unique_lock, or std::lock_guard.
+ */
+class RWLock : public std::recursive_mutex
```

Comment:
Why not make this `ts::shared_mutex` then? Make it an exact implementation of `std::shared_mutex` for future compatibility.